File: ZezeJava/ZezeJava/src/main/java/Zeze/Transaction/Collections/LogOne.java
package Zeze.Transaction.Collections;

import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.IByteBuffer;
import Zeze.Transaction.Bean;
import Zeze.Transaction.Changes;
import Zeze.Transaction.Log;
import Zeze.Transaction.Savepoint;
import Zeze.Util.Task;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/** CollOne 的修改日志：整体换值或收集子 Bean 的修改。 */
@SuppressWarnings("unchecked")
public class LogOne<V extends Bean> extends LogBean {
	private final @NotNull LogOneMeta<V> meta;
	V value;
	@Nullable LogBean logBean;

	public LogOne(Bean belong, int varId, Bean self, @NotNull V value) {
		super(belong, varId, self);
		// 声明类优先（FND7-80）：装子类实例须按声明类建typeId，与读端注册对称
		// （Helper.dependsBean→registerLogOne）；宿主meta未提供时退回运行时类。
		var hostMeta = self instanceof CollOne<?> collOne ? collOne.meta : null;
		meta = hostMeta != null ? (LogOneMeta<V>)hostMeta : LogOneMeta.get((Class<V>)value.getClass());
		this.value = value;
	}

	public LogOne(int varId, @NotNull LogOneMeta<V> meta) {
		super(null, varId, null);
		this.meta = meta; // for decode
	}

	public void setValue(@NotNull V value) {
		this.value = value;
	}

	@Override
	public int getTypeId() {
		return meta.logTypeId;
	}

	@Override
	public @NotNull String getTypeName() {
		return meta.name;
	}

	@Override
	public @NotNull Log beginSavepoint() {
		return new LogOne<>(getBelong(), getVariableId(), getThis(), value);
	}

	@Override
	public void endSavepoint(@NotNull Savepoint currentSp) {
		// 结束保存点，直接覆盖到当前的日志里面即可。
		currentSp.putLog(this);
	}
	// 收集内部的Bean发生了改变。

	@Override
	public void collect(@NotNull Changes changes, @NotNull Bean recent, @NotNull Log vlog) {
		if (logBean == null) {
			logBean = (LogBean)vlog;
			changes.collect(recent, this);
		}
	}

	@Override
	public void commit() {
		if (value != null) // value是否真的可以为null,目前没看到哪里可以让它为null
			((CollOne<V>)getThis()).value = value;
	}

	@Override
	public void encode(@NotNull ByteBuffer bb) {
		if (value != null) { // value是否真的可以为null,目前没看到哪里可以让它为null
			bb.WriteBool(true);
			value.encode(bb);
		} else {
			// 此分支目前不可达：写端构造LogOne时value恒非null（CollOne.createLogBean传getValue、
			// setValue传非null值），仓内也没有"解码后再次编码转发"的路径。
			// 仍按家族标准encodeLogBean写子类判别字节（同LogMap2/LogSortedMap2），保证万一
			// 将来可达（如one-dynamic、日志转发）时线格式对称，不会静默错位。
			bb.WriteBool(false); // value tag
			if (logBean != null) {
				bb.WriteBool(true);
				encodeLogBean(bb, logBean);
			} else
				bb.WriteBool(false);
		}
	}

	@Override
	public void decode(@NotNull IByteBuffer bb) {
		var hasValue = bb.ReadBool();
		if (hasValue) {
			try {
				value = (V)meta.valueFactory.invoke();
			} catch (Throwable e) { // MethodHandle.invoke
				throw Task.forceThrow(e);
			}
			value.decode(bb);
		} else if (bb.ReadBool()) { // hasLogBean。分支不可达性见encode注释；decodeLogBean按判别字节重建子类。
			logBean = decodeLogBean(bb);
		}
	}

	@Override
	public @NotNull String toString() {
		// decode可构造出value==null的LogOne（hasValue=false&&hasLogBean=true，见decode），
		// encode/commit都处理了null，这里不处理的话日志打印级联调用会抛NPE掩盖原始异常。
		return value != null ? value.toString() : "LogOne(logBean=" + logBean + ")";
	}
}

File: Zege/Zeze/Transaction/Collections/LogOne.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Zeze.Serialize;

namespace Zeze.Transaction.Collections
{
    public class LogOne<V> : LogBean
#if !USE_CONFCS
        where V : Bean, new()
#else
        where V : Util.ConfBean, new()
#endif
    {
        public V Value { get; internal set; }
        public LogBean LogBean { get; private set; }

#if !USE_CONFCS
        public void SetValue(V value)
        {
            Value = value;
        }

        internal override Log BeginSavepoint()
        {
            var dup = new LogOne<V>();
            dup.This = This;
            dup.Belong = Belong;
            dup.VariableId = VariableId;
            dup.Value = Value;
            return dup;
        }

        internal override void EndSavepoint(Savepoint currentsp)
        {
            // 结束保存点，直接覆盖到当前的日志里面即可。
            currentsp.PutLog(this);
        }

        // 收集内部的Bean发生了改变。
        public override void Collect(Changes changes, Bean recent, Log vlog)
        {
            if (LogBean == null)
            {
                LogBean = (LogBean)vlog;
                changes.Collect(recent, this);
            }
        }

        public override void Commit()
        {
            if (Value != null)
            {
                ((CollOne<V>)This)._Value = Value;
            }
        }
#endif

        public override void Encode(ByteBuffer bb)
        {
            if (null != Value)
            {
                bb.WriteBool(true);
                Value.Encode(bb);
            }
            else
            {
                bb.WriteBool(false); // Value Tag
                if (null != LogBean)
                {
                    bb.WriteBool(true);
                    LogBean.Encode(bb);
                }
                else
                {
                    bb.WriteBool(false);
                }
            }
        }

        public override void Decode(ByteBuffer bb)
        {
            var hasValue = bb.ReadBool();
            if (hasValue)
            {
                Value = new V();
                Value.Decode(bb);
            }
            else
            {
                var hasLogBean = bb.ReadBool();
                if (hasLogBean)
                {
                    LogBean = new LogBean();
                    LogBean.Decode(bb);
                }
            }
        }

        public override string ToString()
        {
            return Value.ToString();
        }
    }
}
